File: cpp/tests/physics/test_darcy_science.cpp
#include "../test_support/science_test.hpp"
#include <algorithm>
#include <biotransport/core/boundary.hpp>
#include <biotransport/core/mesh/structured_mesh.hpp>
#include <biotransport/physics/fluid_dynamics/darcy_flow.hpp>
#include <cmath>
#include <stdexcept>
#include <string>
#include <vector>

using namespace biotransport;

namespace {

template <typename Exception, typename Callable>
void requireThrows(Callable&& callable, const std::string& context) {
    bool caught = false;
    try {
        callable();
    } catch (const Exception&) {
        caught = true;
    }
    SCIENCE_REQUIRE(caught, context);
}

double harmonicMean(double first, double second) {
    return 2.0 * first * second / (first + second);
}

double layeredPressure(double x, double length, double interface_x, double pressure_left,
                       double pressure_right, double kappa_left, double kappa_right) {
    const double resistance = interface_x / kappa_left + (length - interface_x) / kappa_right;
    const double darcy_flux = (pressure_left - pressure_right) / resistance;
    if (x <= interface_x) {
        return pressure_left - darcy_flux * x / kappa_left;
    }
    const double interface_pressure = pressure_left - darcy_flux * interface_x / kappa_left;
    return interface_pressure - darcy_flux * (x - interface_x) / kappa_right;
}

std::vector<double> layeredMobility(const StructuredMesh& mesh, double interface_x,
                                    double kappa_left, double kappa_right,
                                    bool interface_node_is_left) {
    std::vector<double> kappa(static_cast<std::size_t>(mesh.numNodes()));
    for (int j = 0; j <= mesh.ny(); ++j) {
        for (int i = 0; i <= mesh.nx(); ++i) {
            const double x = mesh.x(i);
            const bool left = x < interface_x || (interface_node_is_left && x == interface_x);
            kappa[static_cast<std::size_t>(mesh.index(i, j))] = left ? kappa_left : kappa_right;
        }
    }
    return kappa;
}

std::vector<double> layeredInitialGuess(const StructuredMesh& mesh, double length,
                                        double interface_x, double pressure_left,
                                        double pressure_right, double kappa_left,
                                        double kappa_right) {
    std::vector<double> pressure(static_cast<std::size_t>(mesh.numNodes()));
    for (int j = 0; j <= mesh.ny(); ++j) {
        for (int i = 0; i <= mesh.nx(); ++i) {
            pressure[static_cast<std::size_t>(mesh.index(i, j))] =
                layeredPressure(mesh.x(i), length, interface_x, pressure_left, pressure_right,
                                kappa_left, kappa_right);
        }
    }
    return pressure;
}

void uniformPressureDropMatchesDarcyLaw() {
    constexpr double length = 0.04;                                                  // m
    constexpr double pressure_left = 3200.0;                                         // Pa
    constexpr double pressure_right = 800.0;                                         // Pa
    constexpr double kappa = 2.5e-10;                                                // m^2/(Pa s)
    constexpr double pressure_gradient = (pressure_right - pressure_left) / length;  // Pa/m
    constexpr double expected_velocity = -kappa * pressure_gradient;                 // m/s

    const StructuredMesh mesh(20, 6, 0.0, length, 0.0, 0.012);
    DarcyFlowSolver solver(mesh, kappa);
    solver.setDirichlet(Boundary::Left, pressure_left)
        .setDirichlet(Boundary::Right, pressure_right)
        .setNeumann(Boundary::Bottom, 0.0)
        .setNeumann(Boundary::Top, 0.0)
        .setOmega(1.65)
        .setTolerance(1.0e-10)
        .setMaxIterations(20000);

    const DarcyFlowResult result = solver.solve();
    double pressure_error = 0.0;
    double velocity_error = 0.0;
    double transverse_velocity = 0.0;
    for (int j = 0; j <= mesh.ny(); ++j) {
        for (int i = 0; i <= mesh.nx(); ++i) {
            const std::size_t index = static_cast<std::size_t>(mesh.index(i, j));
            const double exact_pressure = pressure_left + pressure_gradient * mesh.x(i);
            pressure_error =
                std::max(pressure_error, std::abs(result.pressure[index] - exact_pressure));
            velocity_error =
                std::max(velocity_error, std::abs(result.vx[index] - expected_velocity));
            transverse_velocity = std::max(transverse_velocity, std::abs(result.vy[index]));
        }
    }

    science_test::report("uniform pressure L_inf error", pressure_error, "Pa");
    science_test::report("uniform x-velocity L_inf error", velocity_error, "m/s");
    science_test::report("uniform max |vy|", transverse_velocity, "m/s");
    science_test::report("uniform iterations", static_cast<double>(result.iterations));
    science_test::report("uniform fixed-point defect", result.residual, "Pa");

    SCIENCE_REQUIRE(result.converged, "every returned Darcy result must be converged");
    SCIENCE_REQUIRE(result.residual <= 1.0e-10,
                    "reported pressure defect must satisfy the requested tolerance");
    SCIENCE_REQUIRE(pressure_error < 2.0e-8,
                    "linear pressure must match the analytical Dirichlet solution");
    SCIENCE_REQUIRE(velocity_error < 2.0e-15,
                    "v_x = -kappa dp/dx must hold in SI units on the linear field");
    SCIENCE_REQUIRE(transverse_velocity < 2.0e-15,
                    "zero outward top/bottom gradients must produce zero transverse flow");
}

void outwardGradientSignAndUnitsAreExplicit() {
    constexpr double length = 0.02;                        // m
    constexpr double fixed_pressure = 1200.0;              // Pa
    constexpr double gradient = 25000.0;                   // Pa/m
    constexpr double kappa = 4.0e-10;                      // m^2/(Pa s)
    constexpr double pressure_change = gradient * length;  // Pa
    constexpr double speed = kappa * gradient;             // m/s

    const StructuredMesh mesh(10, 4, 0.0, length, 0.0, 0.008);

    std::vector<double> right_guess(static_cast<std::size_t>(mesh.numNodes()));
    std::vector<double> left_guess(static_cast<std::size_t>(mesh.numNodes()));
    for (int j = 0; j <= mesh.ny(); ++j) {
        for (int i = 0; i <= mesh.nx(); ++i) {
            const std::size_t index = static_cast<std::size_t>(mesh.index(i, j));
            right_guess[index] = fixed_pressure + gradient * mesh.x(i);
            left_guess[index] = fixed_pressure + gradient * (length - mesh.x(i));
        }
    }

    DarcyFlowSolver right_gradient(mesh, kappa);
    right_gradient.setDirichlet(Boundary::Left, fixed_pressure)
        .setNeumann(Boundary::Right, gradient)
        .setNeumann(Boundary::Bottom, 0.0)
        .setNeumann(Boundary::Top, 0.0)
        .setInitialGuess(right_guess)
        .setTolerance(1.0e-11)
        .setMaxIterations(10);
    const DarcyFlowResult right_result = right_gradient.solve();

    DarcyFlowSolver left_gradient(mesh, kappa);
    left_gradient.setNeumann(Boundary::Left, gradient)
        .setDirichlet(Boundary::Right, fixed_pressure)
        .setNeumann(Boundary::Bottom, 0.0)
        .setNeumann(Boundary::Top, 0.0)
        .setInitialGuess(left_guess)
        .setTolerance(1.0e-11)
        .setMaxIterations(10);
    const DarcyFlowResult left_result = left_gradient.solve();

    const std::size_t right_mid = static_cast<std::size_t>(mesh.index(mesh.nx(), mesh.ny() / 2));
    const std::size_t left_mid = static_cast<std::size_t>(mesh.index(0, mesh.ny() / 2));
    SCIENCE_REQUIRE_NEAR(right_result.pressure[right_mid], fixed_pressure + pressure_change,
                         2.0e-11, 0.0, "right-boundary pressure from +dp/dn [Pa/m]");
    SCIENCE_REQUIRE_NEAR(left_result.pressure[left_mid], fixed_pressure + pressure_change, 2.0e-11,
                         0.0, "left-boundary pressure from +dp/dn [Pa/m]");
    SCIENCE_REQUIRE_NEAR(right_result.vx[right_mid], -speed, 2.0e-15, 0.0,
                         "right-boundary Darcy velocity from +dp/dn");
    SCIENCE_REQUIRE_NEAR(left_result.vx[left_mid], speed, 2.0e-15, 0.0,
                         "left-boundary x velocity from +outward dp/dn");

    science_test::report("prescribed outward gradient", gradient, "Pa/m");
    science_test::report("induced pressure change", pressure_change, "Pa");
    science_test::report("right outward Darcy velocity", right_result.vx[right_mid], "m/s");
    science_test::report("left outward Darcy velocity", -left_result.vx[left_mid], "m/s");
}

void layeredMediumConservesNormalFaceFlux() {
    constexpr double length = 1.0;             // m
    constexpr double interface_x = 0.5;        // m
    constexpr double pressure_left = 6000.0;   // Pa
    constexpr double pressure_right = 1000.0;  // Pa
    constexpr double kappa_left = 2.0e-10;     // m^2/(Pa s)
    constexpr double kappa_right = 8.0e-10;    // m^2/(Pa s)
    constexpr double resistance = interface_x / kappa_left + (length - interface_x) / kappa_right;
    constexpr double expected_flux = (pressure_left - pressure_right) / resistance;  // m/s
    constexpr double expected_interface_pressure =
        pressure_left - expected_flux * interface_x / kappa_left;

    // With 31 cells, x=0.5 lies exactly at the face between nodes 15 and 16.
    const StructuredMesh mesh(31, 4, 0.0, length, 0.0, 0.125);
    const std::vector<double> kappa =
        layeredMobility(mesh, interface_x, kappa_left, kappa_right, false);
    DarcyFlowSolver solver(mesh, kappa);
    solver.setDirichlet(Boundary::Left, pressure_left)
        .setDirichlet(Boundary::Right, pressure_right)
        .setNeumann(Boundary::Bottom, 0.0)
        .setNeumann(Boundary::Top, 0.0)
        .setInitialGuess(layeredInitialGuess(mesh, length, interface_x, pressure_left,
                                             pressure_right, kappa_left, kappa_right))
        .setTolerance(1.0e-10)
        .setMaxIterations(100);
    const DarcyFlowResult result = solver.solve();

    const int row = mesh.ny() / 2;
    double min_flux = expected_flux;
    double max_flux = expected_flux;
    double pressure_error = 0.0;
    for (int i = 0; i < mesh.nx(); ++i) {
        const std::size_t west = static_cast<std::size_t>(mesh.index(i, row));
        const std::size_t east = static_cast<std::size_t>(mesh.index(i + 1, row));
        const double face_kappa = harmonicMean(kappa[west], kappa[east]);
        const double face_flux =
            -face_kappa * (result.pressure[east] - result.pressure[west]) / mesh.dx();
        min_flux = std::min(min_flux, face_flux);
        max_flux = std::max(max_flux, face_flux);
    }
    for (int j = 0; j <= mesh.ny(); ++j) {
        for (int i = 0; i <= mesh.nx(); ++i) {
            const std::size_t index = static_cast<std::size_t>(mesh.index(i, j));
            const double exact = layeredPressure(mesh.x(i), length, interface_x, pressure_left,
                                                 pressure_right, kappa_left, kappa_right);
            pressure_error = std::max(pressure_error, std::abs(result.pressure[index] - exact));
        }
    }

    const int left_interface_i = (mesh.nx() - 1) / 2;
    const int right_interface_i = left_interface_i + 1;
    const double left_interface_pressure =
        result.pressure[static_cast<std::size_t>(mesh.index(left_interface_i, row))] -
        expected_flux * (interface_x - mesh.x(left_interface_i)) / kappa_left;
    const double right_interface_pressure =
        result.pressure[static_cast<std::size_t>(mesh.index(right_interface_i, row))] +
        expected_flux * (mesh.x(right_interface_i) - interface_x) / kappa_right;
    const double relative_flux_spread = (max_flux - min_flux) / expected_flux;

    science_test::report("layered analytical face flux", expected_flux, "m/s");
    science_test::report("layered relative face-flux spread", relative_flux_spread);
    science_test::report("layered pressure L_inf error", pressure_error, "Pa");
    science_test::report("layered interface pressure from left", left_interface_pressure, "Pa");
    science_test::report("layered interface pressure from right", right_interface_pressure, "Pa");

    SCIENCE_REQUIRE_NEAR(left_interface_pressure, expected_interface_pressure, 2.0e-7, 0.0,
                         "left piecewise pressure drop to the material interface");
    SCIENCE_REQUIRE_NEAR(right_interface_pressure, expected_interface_pressure, 2.0e-7, 0.0,
                         "right piecewise pressure drop to the material interface");
    SCIENCE_REQUIRE(pressure_error < 2.0e-7,
                    "face-aligned two-material pressure must match the series-resistance solution");
    SCIENCE_REQUIRE(relative_flux_spread < 2.0e-9,
                    "harmonic face mobility must conserve normal Darcy flux across the interface");
}

struct RefinementMeasurement {
    int cells;
    double spacing;
    double pressure_error;
};

RefinementMeasurement nodeAlignedInterfaceError(int cells) {
    constexpr double length = 1.0;
    constexpr double interface_x = 0.5;
    constexpr double pressure_left = 6000.0;
    constexpr double pressure_right = 1000.0;
    constexpr double kappa_left = 2.0e-10;
    constexpr double kappa_right = 8.0e-10;

    const StructuredMesh mesh(cells, 2, 0.0, length, 0.0, 0.125);
    const std::vector<double> kappa =
        layeredMobility(mesh, interface_x, kappa_left, kappa_right, true);
    DarcyFlowSolver solver(mesh, kappa);
    solver.setDirichlet(Boundary::Left, pressure_left)
        .setDirichlet(Boundary::Right, pressure_right)
        .setNeumann(Boundary::Bottom, 0.0)
        .setNeumann(Boundary::Top, 0.0)
        .setInitialGuess(layeredInitialGuess(mesh, length, interface_x, pressure_left,
                                             pressure_right, kappa_left, kappa_right))
        .setOmega(1.6)
        .setTolerance(1.0e-10)
        .setMaxIterations(20000);
    const DarcyFlowResult result = solver.solve();

    double error = 0.0;
    for (int j = 0; j <= mesh.ny(); ++j) {
        for (int i = 0; i <= mesh.nx(); ++i) {
            const std::size_t index = static_cast<std::size_t>(mesh.index(i, j));
            const double exact = layeredPressure(mesh.x(i), length, interface_x, pressure_left,
                                                 pressure_right, kappa_left, kappa_right);
            error = std::max(error, std::abs(result.pressure[index] - exact));
        }
    }
    return {cells, mesh.dx(), error};
}

void discontinuousInterfaceRefinesAtMeasuredFirstOrder() {
    const std::vector<RefinementMeasurement> measurements = {
        nodeAlignedInterfaceError(16),
        nodeAlignedInterfaceError(32),
        nodeAlignedInterfaceError(64),
    };
    const double order_16_to_32 =
        std::log(measurements[0].pressure_error / measurements[1].pressure_error) / std::log(2.0);
    const double order_32_to_64 =
        std::log(measurements[1].pressure_error / measurements[2].pressure_error) / std::log(2.0);

    for (const auto& measurement : measurements) {
        science_test::report("refinement h (N=" + std::to_string(measurement.cells) + ")",
                             measurement.spacing, "m");
        science_test::report(
            "refinement pressure L_inf error (N=" + std::to_string(measurement.cells) + ")",
            measurement.pressure_error, "Pa");
    }
    science_test::report("observed order N=16->32", order_16_to_32);
    science_test::report("observed order N=32->64", order_32_to_64);

    SCIENCE_REQUIRE(measurements[1].pressure_error < measurements[0].pressure_error &&
                        measurements[2].pressure_error < measurements[1].pressure_error,
                    "the represented interface error must decrease monotonically under refinement");
    SCIENCE_REQUIRE(order_16_to_32 > 0.8 && order_16_to_32 < 1.2,
                    "the measured discontinuous, node-aligned interface rate is first order");
    SCIENCE_REQUIRE(order_32_to_64 > 0.8 && order_32_to_64 < 1.2,
                    "the measured discontinuous, node-aligned interface rate is first order");
}

void singularAndUnconvergedProblemsFailLoudly() {
    const StructuredMesh mesh(8, 4, 0.0, 1.0, 0.0, 0.5);

    DarcyFlowSolver unanchored(mesh, 1.0e-10);
    requireThrows<std::invalid_argument>(
        [&] { (void)unanchored.solve(); },
        "an all-Neumann pressure system must reject its missing gauge");

    DarcyFlowSolver exhausted(mesh, 1.0e-10);
    exhausted.setDirichlet(Boundary::Left, 2000.0)
        .setDirichlet(Boundary::Right, 1000.0)
        .setNeumann(Boundary::Bottom, 0.0)
        .setNeumann(Boundary::Top, 0.0)
        .setTolerance(1.0e-15)
        .setMaxIterations(1);
    requireThrows<std::runtime_error>(
        [&] { (void)exhausted.solve(); },
        "iteration exhaustion must throw instead of returning a field");
}

}  // namespace

int main() {
    return science_test::runSuite(
        "Darcy flow analytical and failure verification",
        {{"uniform analytical pressure and velocity", uniformPressureDropMatchesDarcyLaw},
         {"outward-gradient sign and SI units", outwardGradientSignAndUnitsAreExplicit},
         {"two-material pressure drop and flux continuity", layeredMediumConservesNormalFaceFlux},
         {"honest discontinuous-interface refinement",
          discontinuousInterfaceRefinesAtMeasuredFirstOrder},
         {"gauge and convergence failures", singularAndUnconvergedProblemsFailLoudly}});
}

File: python/tests/test_darcy_science.py
"""Independent science checks for the public Darcy-flow interface."""

from __future__ import annotations

from collections.abc import Callable
import math

import numpy as np
import pytest

import biotransport as bt


LENGTH_M = 1.0
INTERFACE_M = 0.5
PRESSURE_LEFT_PA = 6000.0
PRESSURE_RIGHT_PA = 1000.0
KAPPA_LEFT_M2_PER_PA_S = 2.0e-10
KAPPA_RIGHT_M2_PER_PA_S = 8.0e-10


def _layered_flux() -> float:
    resistance = (
        INTERFACE_M / KAPPA_LEFT_M2_PER_PA_S
        + (LENGTH_M - INTERFACE_M) / KAPPA_RIGHT_M2_PER_PA_S
    )
    return (PRESSURE_LEFT_PA - PRESSURE_RIGHT_PA) / resistance


def _layered_pressure(x_m: float) -> float:
    flux_m_per_s = _layered_flux()
    if x_m <= INTERFACE_M:
        return PRESSURE_LEFT_PA - flux_m_per_s * x_m / KAPPA_LEFT_M2_PER_PA_S
    pressure_at_interface = (
        PRESSURE_LEFT_PA - flux_m_per_s * INTERFACE_M / KAPPA_LEFT_M2_PER_PA_S
    )
    return pressure_at_interface - (
        flux_m_per_s * (x_m - INTERFACE_M) / KAPPA_RIGHT_M2_PER_PA_S
    )


def _field(mesh: bt.StructuredMesh, value_at_x: Callable[[float], float]) -> np.ndarray:
    values = np.empty(mesh.num_nodes(), dtype=np.float64)
    for j in range(mesh.ny() + 1):
        for i in range(mesh.nx() + 1):
            values[mesh.index(i, j)] = value_at_x(mesh.x(i))
    return values


def _layered_mobility(
    mesh: bt.StructuredMesh, *, interface_node_is_left: bool
) -> np.ndarray:
    def mobility(x_m: float) -> float:
        is_left = x_m < INTERFACE_M or (interface_node_is_left and x_m == INTERFACE_M)
        return KAPPA_LEFT_M2_PER_PA_S if is_left else KAPPA_RIGHT_M2_PER_PA_S

    return _field(mesh, mobility)


def _solve_node_aligned_interface(cells: int) -> tuple[float, float]:
    mesh = bt.StructuredMesh(cells, 2, 0.0, LENGTH_M, 0.0, 0.125)
    kappa = _layered_mobility(mesh, interface_node_is_left=True)
    solver = bt.DarcyFlowSolver(mesh, kappa.tolist())
    solver.set_dirichlet(bt.Boundary.Left, PRESSURE_LEFT_PA)
    solver.set_dirichlet(bt.Boundary.Right, PRESSURE_RIGHT_PA)
    solver.set_outward_pressure_gradient(bt.Boundary.Bottom, 0.0)
    solver.set_outward_pressure_gradient(bt.Boundary.Top, 0.0)
    solver.set_initial_guess(_field(mesh, _layered_pressure).tolist())
    solver.set_omega(1.6)
    solver.set_tolerance(1.0e-10)
    solver.set_max_iterations(20_000)
    pressure = np.asarray(solver.solve().pressure())
    exact = _field(mesh, _layered_pressure)
    return mesh.dx(), float(np.max(np.abs(pressure - exact)))


def test_uniform_dirichlet_drop_matches_pressure_and_velocity_solution() -> None:
    length_m = 0.04
    pressure_left_pa = 3200.0
    pressure_right_pa = 800.0
    kappa_m2_per_pa_s = 2.5e-10
    gradient_pa_per_m = (pressure_right_pa - pressure_left_pa) / length_m
    velocity_m_per_s = -kappa_m2_per_pa_s * gradient_pa_per_m

    mesh = bt.StructuredMesh(20, 6, 0.0, length_m, 0.0, 0.012)
    solver = bt.DarcyFlowSolver(mesh, kappa_m2_per_pa_s)
    solver.set_dirichlet(bt.Boundary.Left, pressure_left_pa)
    solver.set_dirichlet(bt.Boundary.Right, pressure_right_pa)
    solver.set_outward_pressure_gradient(bt.Boundary.Bottom, 0.0)
    solver.set_outward_pressure_gradient(bt.Boundary.Top, 0.0)
    solver.set_omega(1.65)
    solver.set_tolerance(1.0e-10)
    solver.set_max_iterations(20_000)
    result = solver.solve()

    exact_pressure = _field(
        mesh, lambda x_m: pressure_left_pa + gradient_pa_per_m * x_m
    )
    assert result.converged
    assert result.residual <= 1.0e-10
    assert np.max(np.abs(np.asarray(result.pressure()) - exact_pressure)) < 2.0e-8
    assert np.max(np.abs(np.asarray(result.vx()) - velocity_m_per_s)) < 2.0e-15
    assert np.max(np.abs(np.asarray(result.vy()))) < 2.0e-15


def test_neumann_value_is_outward_pressure_gradient_in_pa_per_m() -> None:
    length_m = 0.02
    fixed_pressure_pa = 1200.0
    gradient_pa_per_m = 25_000.0
    kappa_m2_per_pa_s = 4.0e-10
    expected_pressure_pa = fixed_pressure_pa + gradient_pa_per_m * length_m
    speed_m_per_s = kappa_m2_per_pa_s * gradient_pa_per_m
    mesh = bt.StructuredMesh(10, 4, 0.0, length_m, 0.0, 0.008)

    right_solver = bt.DarcyFlowSolver(mesh, kappa_m2_per_pa_s)
    right_solver.set_dirichlet(bt.Boundary.Left, fixed_pressure_pa)
    right_solver.set_outward_pressure_gradient(bt.Boundary.Right, gradient_pa_per_m)
    right_solver.set_outward_pressure_gradient(bt.Boundary.Bottom, 0.0)
    right_solver.set_outward_pressure_gradient(bt.Boundary.Top, 0.0)
    right_solver.set_initial_guess(
        _field(mesh, lambda x_m: fixed_pressure_pa + gradient_pa_per_m * x_m).tolist()
    )
    right_solver.set_tolerance(1.0e-11)
    right_solver.set_max_iterations(10)
    right = right_solver.solve()

    left_solver = bt.DarcyFlowSolver(mesh, kappa_m2_per_pa_s)
    left_solver.set_outward_pressure_gradient(bt.Boundary.Left, gradient_pa_per_m)
    left_solver.set_dirichlet(bt.Boundary.Right, fixed_pressure_pa)
    left_solver.set_outward_pressure_gradient(bt.Boundary.Bottom, 0.0)
    left_solver.set_outward_pressure_gradient(bt.Boundary.Top, 0.0)
    left_solver.set_initial_guess(
        _field(
            mesh,
            lambda x_m: fixed_pressure_pa + gradient_pa_per_m * (length_m - x_m),
        ).tolist()
    )
    left_solver.set_tolerance(1.0e-11)
    left_solver.set_max_iterations(10)
    left = left_solver.solve()

    right_mid = mesh.index(mesh.nx(), mesh.ny() // 2)
    left_mid = mesh.index(0, mesh.ny() // 2)
    assert np.asarray(right.pressure())[right_mid] == pytest.approx(
        expected_pressure_pa, abs=2.0e-11
    )
    assert np.asarray(left.pressure())[left_mid] == pytest.approx(
        expected_pressure_pa, abs=2.0e-11
    )
    assert np.asarray(right.vx())[right_mid] == pytest.approx(
        -speed_m_per_s, abs=2.0e-15
    )
    assert np.asarray(left.vx())[left_mid] == pytest.approx(speed_m_per_s, abs=2.0e-15)


def test_face_aligned_two_material_solution_has_continuous_normal_flux() -> None:
    mesh = bt.StructuredMesh(31, 4, 0.0, LENGTH_M, 0.0, 0.125)
    kappa = _layered_mobility(mesh, interface_node_is_left=False)
    solver = bt.DarcyFlowSolver(mesh, kappa.tolist())
    solver.set_dirichlet(bt.Boundary.Left, PRESSURE_LEFT_PA)
    solver.set_dirichlet(bt.Boundary.Right, PRESSURE_RIGHT_PA)
    solver.set_outward_pressure_gradient(bt.Boundary.Bottom, 0.0)
    solver.set_outward_pressure_gradient(bt.Boundary.Top, 0.0)
    solver.set_initial_guess(_field(mesh, _layered_pressure).tolist())
    solver.set_tolerance(1.0e-10)
    solver.set_max_iterations(100)
    result = solver.solve()
    pressure = np.asarray(result.pressure())
    expected_pressure = _field(mesh, _layered_pressure)

    row = mesh.ny() // 2
    face_fluxes = []
    for i in range(mesh.nx()):
        west = mesh.index(i, row)
        east = mesh.index(i + 1, row)
        face_kappa = 2.0 * kappa[west] * kappa[east] / (kappa[west] + kappa[east])
        face_fluxes.append(-face_kappa * (pressure[east] - pressure[west]) / mesh.dx())
    face_fluxes_array = np.asarray(face_fluxes)
    expected_flux = _layered_flux()
    relative_flux_spread = float(np.ptp(face_fluxes_array) / expected_flux)

    left_i = (mesh.nx() - 1) // 2
    right_i = left_i + 1
    interface_pressure_from_left = pressure[mesh.index(left_i, row)] - (
        expected_flux * (INTERFACE_M - mesh.x(left_i)) / KAPPA_LEFT_M2_PER_PA_S
    )
    interface_pressure_from_right = pressure[mesh.index(right_i, row)] + (
        expected_flux * (mesh.x(right_i) - INTERFACE_M) / KAPPA_RIGHT_M2_PER_PA_S
    )
    expected_interface_pressure = (
        PRESSURE_LEFT_PA - expected_flux * INTERFACE_M / KAPPA_LEFT_M2_PER_PA_S
    )

    assert np.max(np.abs(pressure - expected_pressure)) < 2.0e-7
    assert relative_flux_spread < 2.0e-9
    assert interface_pressure_from_left == pytest.approx(
        expected_interface_pressure, abs=2.0e-7
    )
    assert interface_pressure_from_right == pytest.approx(
        expected_interface_pressure, abs=2.0e-7
    )


def test_node_aligned_discontinuous_interface_has_measured_first_order_error() -> None:
    measurements = [_solve_node_aligned_interface(cells) for cells in (16, 32, 64)]
    errors = [measurement[1] for measurement in measurements]
    orders = [math.log(errors[i] / errors[i + 1], 2.0) for i in range(2)]

    assert errors[2] < errors[1] < errors[0]
    assert all(0.8 < order < 1.2 for order in orders)


def test_unanchored_and_forced_unconverged_systems_raise() -> None:
    mesh = bt.StructuredMesh(8, 4, 0.0, 1.0, 0.0, 0.5)

    with pytest.raises(ValueError, match="unanchored"):
        bt.DarcyFlowSolver(mesh, 1.0e-10).solve()

    exhausted = bt.DarcyFlowSolver(mesh, 1.0e-10)
    exhausted.set_dirichlet(bt.Boundary.Left, 2000.0)
    exhausted.set_dirichlet(bt.Boundary.Right, 1000.0)
    exhausted.set_outward_pressure_gradient(bt.Boundary.Bottom, 0.0)
    exhausted.set_outward_pressure_gradient(bt.Boundary.Top, 0.0)
    exhausted.set_tolerance(1.0e-15)
    exhausted.set_max_iterations(1)
    with pytest.raises(RuntimeError, match="did not converge"):
        exhausted.solve()
